### PR TITLE
feat: display sockets in process tree

### PR DESCRIPTION
**Pull request by snprajwal**
This is now supported with the `--sockets` flag.

### CODE REVIEW

**Comment by snprajwal** [comment]
Is Codecov broken? It seems to not update the coverage, or report the status via CI anymore

**Comment by rst0git** [comment]
> Is Codecov broken? It seems to not update the coverage, or report the status via CI anymore

It seems to work:

- https://github.com/checkpoint-restore/checkpointctl/actions/runs/5820452857/job/15782983383?pr=94
- https://app.codecov.io/gh/checkpoint-restore/checkpointctl/pull/94

File: tree.go [hunk 1]
```diff
@@ -173,6 +183,35 @@ func addFdsToTree(tree treeprint.Tree, fds []*crit.Fd) {
...
+				data = fmt.Sprintf("%s:%d -> %s:%d [%s]", socket.SrcAddr, socket.SrcPort, socket.DestAddr, socket.DestPort, socket.State)
+			case "UNIXSK":
+				data = fmt.Sprintf("%s [%s]", socket.SrcAddr, socket.State)
+			case "PACKETSK":
+				data = fmt.Sprintf("↑ %s ↓ %s", socket.SendBuf, socket.RecvBuf)
+			case "NETLINKSK":
+				data = fmt.Sprintf("↑ %s ↓ %s [%s]", socket.SendBuf, socket.RecvBuf, socket.State)
+			}
+
+			node.AddMetaBranch(fmt.Sprintf("%s (%s)", socket.Type, socket.Protocol), data)
```
Comment (rst0git) [inline tree.go:210]: `socket.Protocol` is not set for Unix sockets.
```
jira
├── Image: docker.io/atlassian/jira-software:latest
├── ID: 3f731c3a3951e25cc6509a3d49d7e8b799dc14dec4a96abb95e1ece55fb1692c
├── Runtime: crun
├── Created: 2023-08-11T10:24:48+01:00
├── Engine: Podman
├── Checkpoint size: 495.1 MiB
├── Root FS diff size: 74.5 KiB
└── Process tree
    └── [1]  tini
        └── [2]  java
            ├── [STREAM ()]   [ESTABLISHED]
            ├── [STREAM (TCP)]  :::8080 -> :::0 [LISTEN]
            ├── [STREAM ()]   [ESTABLISHED]
            └── [STREAM (TCP)]  7f00:1:::8005 -> :::0 [LISTEN]
```

File: tree.go [hunk 2]
```diff
@@ -173,6 +183,35 @@ func addFdsToTree(tree treeprint.Tree, fds []*crit.Fd) {
...
+		// Hence, skip all associated sockets.
+		if node == nil {
+			continue
+		}
+
+		var data string
+		for _, socket := range sk.Sockets {
+			switch socket.FdType {
+			case "INETSK":
+				data = fmt.Sprintf("%s:%d -> %s:%d [%s]", socket.SrcAddr, socket.SrcPort, socket.DestAddr, socket.DestPort, socket.State)
```
Comment (rst0git) [inline tree.go:201]: Should we display buffer size here?
```Go
data = fmt.Sprintf("%s:%d -> %s:%d [%s] (↑ %s ↓ %s)", socket.SrcAddr, socket.SrcPort, socket.DestAddr,
            socket.DestPort, socket.State, socket.SendBuf, socket.RecvBuf)
```
```
iperf-client
├── Image: docker.io/networkstatic/iperf3:latest
├── ID: 9a6e7537045cf536d00fc28072e092db92f7339d84fc1481172cbdda37e8aebb
├── Runtime: crun
├── Created: 2023-08-11T12:38:31+01:00
├── Engine: Podman
├── Checkpoint size: 393.3 KiB
└── Process tree
    └── [1]  iperf3
        ├── [STREAM (TCP)]  10.88.0.48:37930 -> 10.88.0.46:5201 [ESTABLISHED] (↑ 85.0 KB ↓ 128.0 KB)
        └── [DGRAM (UDP)]  10.88.0.48:57071 -> 10.88.0.46:5201 [ESTABLISHED] (↑ 256.0 MB ↓ 256.0 MB)
```

File: tree.go (hunk 1 above)
Comment (snprajwal) [reply tree.go:210]: Oops my bad, will get rid of that

File: tree.go [hunk 3]
```diff
@@ -173,6 +183,52 @@ func addFdsToTree(tree treeprint.Tree, fds []*crit.Fd) {
...
+		}
+
+		var data string
+		for _, socket := range sk.Sockets {
+			switch socket.FdType {
+			case "UNIXSK":
+				// UNIX sockets do not have a protocol assigned.
+				// Hence, the meta value for the socket is just
+				// the socket type.
+				node.AddMetaBranch(socket.Type, socket.SrcAddr)
```
Comment (rst0git) [inline tree.go:204]:
```Suggestion
                               data := socket.SrcAddr
                               if len(data) == 0 {
                                       // An abstract socket address
                                       data = "@"
                               }
                               node.AddMetaBranch(fmt.Sprintf("UNIX (%s)", socket.Type), data)
```

File: tree.go [hunk 4]
```diff
@@ -173,6 +183,52 @@ func addFdsToTree(tree treeprint.Tree, fds []*crit.Fd) {
...
+					state,
+					socket.SendBuf, socket.RecvBuf,
+				)
+			case "PACKETSK":
+				data = fmt.Sprintf("↑ %s ↓ %s", socket.SendBuf, socket.RecvBuf)
+			case "NETLINKSK":
+				data = fmt.Sprintf("↑ %s ↓ %s", socket.SendBuf, socket.RecvBuf)
+			}
+
+			node.AddMetaBranch(fmt.Sprintf("%s (%s)", socket.Type, socket.Protocol), data)
```
Comment (rst0git) [inline tree.go:227]:
```diff
                for _, socket := range sk.Sockets {
+                       var protocol = socket.Protocol
                        switch socket.FdType {
                        case "UNIXSK":
                                // UNIX sockets do not have a protocol assigned.
@@ -204,18 +205,16 @@ func addSkToTree(tree treeprint.Tree, sks []*crit.Sk) {
                                node.AddMetaBranch(socket.Type, socket.SrcAddr)
                                continue
                        case "INETSK":
-                               var state string
-                               if socket.Protocol == "TCP" {
+                               if protocol == "TCP" {
                                        // Create state string like "[ESTABLISHED] ".
                                        // The trailing space is used to pad the state
                                        // before the buffer size is printed.
-                                       state = "[" + socket.State + "] "
+                                       protocol = fmt.Sprintf("%s (%s)", protocol, socket.State)
                                }
                                data = fmt.Sprintf(
-                                       "%s:%d -> %s:%d %s(↑ %s ↓ %s)",
+                                       "%s:%d -> %s:%d (↑ %s ↓ %s)",
                                        socket.SrcAddr, socket.SrcPort,
                                        socket.DestAddr, socket.DestPort,
-                                       state,
                                        socket.SendBuf, socket.RecvBuf,
                                )
                        case "PACKETSK":
@@ -224,7 +223,7 @@ func addSkToTree(tree treeprint.Tree, sks []*crit.Sk) {
                                data = fmt.Sprintf("↑ %s ↓ %s", socket.SendBuf, socket.RecvBuf)
                        }
 
-                       node.AddMetaBranch(fmt.Sprintf("%s (%s)", socket.Type, socket.Protocol), data)
+                       node.AddMetaBranch(protocol, data)
```
Examples:
```console
iperf-client
├── Image: docker.io/networkstatic/iperf3:latest
├── ID: 6312a4ee1c503de61e25cf5c27468894d99c0a97614919ab27f3ae4d765aa1b2
├── Runtime: crun
├── Created: 2023-08-14T00:52:42+01:00
├── Engine: Podman
├── Checkpoint size: 385.3 KiB
└── Process tree
    └── [1]  iperf3
        ├── [TCP (ESTABLISHED)]  10.88.0.8:43720 -> 10.88.0.7:5201 (↑ 85.0 KB ↓ 128.0 KB)
        └── [UDP]  10.88.0.8:43468 -> 10.88.0.7:5201 (↑ 256.0 MB ↓ 256.0 MB)
```
```console
jira
├── Image: docker.io/atlassian/jira-software:latest
├── ID: 6fdd46a060e75c7ab5b50e4d3d4bd1a974a5d94acf1ea86e09970fe73119a9b9
├── Runtime: crun
├── Created: 2023-08-14T00:21:34+01:00
├── Engine: Podman
├── Checkpoint size: 605.0 MiB
├── Root FS diff size: 176.5 KiB
└── Process tree
    └── [1]  tini
        └── [2]  java
            ├── [TCP (ESTABLISHED)]  a58:5:::8080 -> a58:1:::42584 (↑ 85.0 KB ↓ 128.0 KB)
            ├── [UNIX (STREAM)]  @
            ├── [TCP (LISTEN)]  :::8080 -> :::0 (↑ 16.0 KB ↓ 128.0 KB)
            ├── [UNIX (STREAM)]  @
            ├── [TCP (LISTEN)]  7f00:1:::8005 -> :::0 (↑ 16.0 KB ↓ 128.0 KB)
            ├── [TCP (ESTABLISHED)]  a58:5:::8080 -> a58:1:::42592 (↑ 85.0 KB ↓ 128.0 KB)
            ├── [TCP (ESTABLISHED)]  a58:5:::8080 -> a58:1:::42608 (↑ 85.0 KB ↓ 128.0 KB)
            ├── [TCP (ESTABLISHED)]  a58:5:::8080 -> a58:1:::42610 (↑ 85.0 KB ↓ 128.0 KB)
            └── [TCP (ESTABLISHED)]  a58:5:::8080 -> a58:1:::42614 (↑ 85.0 KB ↓ 128.0 KB)
```

```console
postgres
├── Image: docker.io/library/postgres:latest
├── ID: b5a13117175120f2dc8f5569ee8200521a65731784a67c94c329fdde3465adb5
├── Runtime: crun
├── Created: 2023-08-14T00:32:15+01:00
├── Engine: Podman
├── Checkpoint size: 25.0 MiB
├── Root FS diff size: 2.0 KiB
└── Process tree
    └── [1]  postgres
        ├── [60]  postgres
        ├── [55]  postgres
        ├── [56]  postgres
        ├── [58]  postgres
        ├── [59]  postgres
        ├── [TCP (LISTEN)]  0.0.0.0:5432 -> 0.0.0.0:0 (↑ 16.0 KB ↓ 128.0 KB)
        ├── [TCP (LISTEN)]  :::5432 -> :::0 (↑ 16.0 KB ↓ 128.0 KB)
        └── [UNIX (STREAM)]  /var/run/postgresql/.s.PGSQL.5432
```

**Pull request merged**